### ISSUE TITLE
[selectors] querySelector() with quote and/or square bracket missing on right

### ISSUE THREAD

**Issue by ewilligers** [w3c/csswg-drafts#1933]
(From Chrome [bug](https://bugs.chromium.org/p/chromium/issues/detail?id=779942))

The call
```
// double quote and square bracket on the right side was missing.
document.querySelector('meta[name="description');
```
succeeds on Chrome/Edge/Firefox and fails on Safari.

Does the spec specify which behavior is correct?

We should have a web platform test. Apart from a DOM [test](https://github.com/w3c/web-platform-tests/blob/master/dom/nodes/ParentNode-querySelector-All.html
), querySelector[All] doesn't appear to have tests.

**Comment by Loirooriol** [comment]
The example can be simplified to `document.querySelector('[n');`

As expected, the spec says it should fail:

1. [`querySelector`](https://dom.spec.whatwg.org/#dom-parentnode-queryselector) does a [scope-match a selectors string](https://dom.spec.whatwg.org/#dom-parentnode-queryselector).
2. [Scope-match a selectors string](https://dom.spec.whatwg.org/#dom-parentnode-queryselector) [parses a selector](https://drafts.csswg.org/selectors-4/#parse-a-selector).
3. [Parse a selector](https://drafts.csswg.org/selectors-4/#parse-a-selector) parses as a [`<selector-list>`](https://drafts.csswg.org/selectors-4/#typedef-selector-list).
4. A `[` in a [`<selector-list>`](https://drafts.csswg.org/selectors-4/#typedef-selector-list) begins an [`<attribute-selector>`](https://drafts.csswg.org/selectors-4/#typedef-attribute-selector), which must be closed with `]`.
5. Therefore the string does not match the grammar, and [parse a selector](https://drafts.csswg.org/selectors-4/#parse-a-selector) returns failure.
6. Therefore [scope-match a selectors string](https://dom.spec.whatwg.org/#scope-match-a-selectors-string) throws a "SyntaxError" DOMException.

**Comment by phistuck** [comment]
I guess non-Safari count `[...` as part of the "etc" in [CSS Syntax Module Level 3 - Error Handling](https://www.w3.org/TR/css-syntax-3/#error-handling) -
> If the stylesheet ends while any rule, declaration, function, string, etc. are still open, everything is automatically closed. This doesn’t make them invalid, though they may be incomplete and thus thrown away when they are verified against their grammar.

**Comment by tabatkins** [comment]
Yeah, this is accepted just fine. The grammar doesn't work over text, it works over post-parsing token structures. In particular, `[]`-blocks get "auto-closed" by EOF (or end-of-string, in this case) - you can't tell whether or not there was a `]` character, and in fact it doesn't matter; all that matters is that there *is* a `[]`-block.  ([See the last paragraph in Syntax ch8.](https://drafts.csswg.org/css-syntax/#rule-defs))

In other words, as far as the grammar is concerned, the stuff it's checking over is:

```
IDENT(meta)
[]-BLOCK
    IDENT(name)
    DELIM(=)
    STRING("description")
```

**Comment by dbaron** [comment]
Also see the "Unexpected end of style sheet" subsection of the [Rules for handling parsing errors](https://www.w3.org/TR/2011/REC-CSS2-20110607/syndata.html#parsing-errors) in CSS 2.1.

**Comment by Loirooriol** [comment]
> The grammar doesn't work over text, it works over post-parsing token structures. 

This is an important point which I think should me mentioned in [parse a selector](https://drafts.csswg.org/selectors-4/#parse-a-selector). Note it explicitly receives a string, not a list of tokens. The algorithm should link the appropriate CSS Syntax section which explains how to tokenize the string.

**Comment by tabatkins** [comment]
The Syntax spec [defines an API hook for this](https://drafts.csswg.org/css-syntax/#parse-grammar). In a Bikeshedded spec, linking this is no harder than writing `[=CSS/parse=]`.

**Comment by gsnedders** [comment]
see also https://github.com/w3c/csswg-drafts/issues/492

**Comment by lestertraumatic** [comment]
> The example can be simplified to `document.querySelector('[n');`
> 
> As expected, the spec says it should fail:
> 
> 1. [`querySelector`](https://dom.spec.whatwg.org/#dom-parentnode-queryselector) does a [scope-match a selectors string](https://dom.spec.whatwg.org/#dom-parentnode-queryselector).
> 2. [Scope-match a selectors string](https://dom.spec.whatwg.org/#dom-parentnode-queryselector) [parses a selector](https://drafts.csswg.org/selectors-4/#parse-a-selector).
> 3. [Parse a selector](https://drafts.csswg.org/selectors-4/#parse-a-selector) parses as a [`<selector-list>`](https://drafts.csswg.org/selectors-4/#typedef-selector-list).
> 4. A `[` in a [`<selector-list>`](https://drafts.csswg.org/selectors-4/#typedef-selector-list) begins an [`<attribute-selector>`](https://drafts.csswg.org/selectors-4/#typedef-attribute-selector), which must be closed with `]`.
> 5. Therefore the string does not match the grammar, and [parse a selector](https://drafts.csswg.org/selectors-4/#parse-a-selector) returns failure.
> 6. Therefore [scope-match a selectors string](https://dom.spec.whatwg.org/#scope-match-a-selectors-string) throws a "SyntaxError" DOMException.





https://stompjs.github.io/decoration10/
